lib/Conversion/SeqToSV/SeqToSV.cpp

```diff
@@ -528,60 +561,72 @@ void SeqToSVPass::runOnOperation() {
-
-  if (hasMemRandomization)
-    emitGuard("RANDOMIZE",
-              [&]() { emitGuardedDefine("RANDOMIZE_MEM_INIT", "RANDOMIZE"); });
-
-  if (hasRegRandomization)
-    emitGuard("RANDOMIZE",
-              [&]() { emitGuardedDefine("RANDOMIZE_REG_INIT", "RANDOMIZE"); });
-
-  b.create<sv::VerbatimOp>(
-      "\n// RANDOM may be set to an expression that produces a 32-bit "
-      "random unsigned value.");
-  emitGuardedDefine("RANDOM", "RANDOM", StringRef(), "$random");
-
-  b.create<sv::VerbatimOp>(
-      "\n// Users can define INIT_RANDOM as general code that gets "
-      "injected "
-      "into the\n// initializer block for modules with registers.");
-  emitGuardedDefine("INIT_RANDOM", "INIT_RANDOM", StringRef(), "");
-
-  b.create<sv::VerbatimOp>(
-      "\n// If using random initialization, you can also define "
-      "RANDOMIZE_DELAY to\n// customize the delay used, otherwise 0.002 "
-      "is used.");
-  emitGuardedDefine("RANDOMIZE_DELAY", "RANDOMIZE_DELAY", StringRef(), "0.002");
-
-  b.create<sv::VerbatimOp>(
-      "\n// Define INIT_RANDOM_PROLOG_ for use in our modules below.");
-  emitGuard("INIT_RANDOM_PROLOG_", [&]() {
-    b.create<sv::IfDefOp>(
-        "RANDOMIZE",
-        [&]() {
-          emitGuardedDefine("VERILATOR", "INIT_RANDOM_PROLOG_", "`INIT_RANDOM",
-                            "`INIT_RANDOM #`RANDOMIZE_DELAY begin end");
-        },
-        [&]() { b.create<sv::MacroDefOp>("INIT_RANDOM_PROLOG_", ""); });
+  b.create<emit::FragmentOp>(randomInitFragmentName.getAttr(), [&] {
+    b.create<sv::VerbatimOp>(
+        "// Standard header to adapt well known macros for "
+        "register randomization.");
+
+    b.create<sv::VerbatimOp>(
+        "\n// RANDOM may be set to an expression that produces a 32-bit "
+        "random unsigned value.");
+    emitGuardedDefine("RANDOM", "RANDOM", StringRef(), "$random");
+
+    b.create<sv::VerbatimOp>(
+        "\n// Users can define INIT_RANDOM as general code that gets "
+        "injected "
+        "into the\n// initializer block for modules with registers.");
+    emitGuardedDefine("INIT_RANDOM", "INIT_RANDOM", StringRef(), "");
+
+    b.create<sv::VerbatimOp>(
+        "\n// If using random initialization, you can also define "
+        "RANDOMIZE_DELAY to\n// customize the delay used, otherwise 0.002 "
+        "is used.");
+    emitGuardedDefine("RANDOMIZE_DELAY", "RANDOMIZE_DELAY", StringRef(),
+                      "0.002");
+
+    b.create<sv::VerbatimOp>(
+        "\n// Define INIT_RANDOM_PROLOG_ for use in our modules below.");
+    emitGuard("INIT_RANDOM_PROLOG_", [&]() {
+      b.create<sv::IfDefOp>(
+          "RANDOMIZE",
+          [&]() {
+            emitGuardedDefine("VERILATOR", "INIT_RANDOM_PROLOG_",
+                              "`INIT_RANDOM",
+                              "`INIT_RANDOM #`RANDOMIZE_DELAY begin end");
+          },
+          [&]() { b.create<sv::MacroDefOp>("INIT_RANDOM_PROLOG_", ""); });
+    });
```

Comment:
Should this check if the fragments already exist, and only add them if they don't? Otherwise you can't pass MLIR that already has run through SeqToSV through firtool.

Reply:
That was broken, will stay broken.